src/Compiler/Utilities/Caches.fs

```diff
@@ -47,112 +48,80 @@ type CachedEntity<'Key, 'Value> =
 
-        listener.SetMeasurementEventCallback<int64>(fun k v _ _ -> Interlocked.Add(readings.GetOrAdd(k.Name, ref 0L), v) |> ignore)
-        listener.Start()
+        builder.ToImmutable()
 
-    member this.Dispose() = listener.Dispose()
+    let incr key v =
+        Interlocked.Add(totals[key], v) |> ignore
 
-    member val CacheId = cacheId
+    let total key = totals[key].Value
 
-    static member val Meter = meter
+    let mutable ratio = Double.NaN
 
-    member val RecentStats = "-" with get, set
-
-    member this.TryUpdateStats(clearCounts) =
-        let ratio =
-            try
-                float readings["hits"].Value
-                / float (readings["hits"].Value + readings["misses"].Value)
-                * 100.0
-            with _ ->
-                Double.NaN
-
-        let stats =
-            [
-                for name in readings.Keys do
-                    let v = readings[name].Value
-
-                    if v > 0 then
-                        $"{name}: {v}"
-            ]
-            |> String.concat ", "
-            |> sprintf "%s | hit ratio: %s %s" this.CacheId (if Double.IsNaN(ratio) then "-" else $"%.1f{ratio}%%")
-
-        if clearCounts then
-            for r in readings.Values do
-                Interlocked.Exchange(r, 0L) |> ignore
-
-        if stats <> this.RecentStats then
-            this.RecentStats <- stats
-            true
-        else
-            false
+    let updateRatio () =
+        ratio <- float (total hits) / float (total hits + total misses)
 
-    // TODO: Should return a Map, not a string
-    static member GetStats(cacheId) =
-        observedCaches[cacheId].TryUpdateStats(false) |> ignore
-        observedCaches[cacheId].RecentStats
+    let listener = new MeterListener()
 
-    static member GetStatsUpdateForAllCaches(clearCounts) =
-        [
-            for i in observedCaches.Values do
-                if i.TryUpdateStats(clearCounts) then
-                    i.RecentStats
-        ]
-        |> String.concat "\n"
+    let startListening () =
+        for i in allCouinters do
+            listener.EnableMeasurementEvents i
 
-    static member AddInstrumentation(cacheId) =
-        if observedCaches.ContainsKey cacheId then
-            invalidArg "cacheId" $"cache with name {cacheId} already exists"
+        listener.SetMeasurementEventCallback(fun instrument v _ _ ->
+            incr instrument v
 
-        observedCaches[cacheId] <- new CacheMetrics(cacheId)
+            if instrument = hits || instrument = misses then
+                updateRatio ())
 
-    static member RemoveInstrumentation(cacheId) =
-        observedCaches[cacheId].Dispose()
-        observedCaches.TryRemove(cacheId) |> ignore
+        listener.Start()
 
-// Creates and after reclaiming holds entities for reuse.
-// More than totalCapacity can be created, but it will hold for reuse at most totalCapacity.
-type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
```

Comment:
Pooling is just removed for now. If there is impact on allocations, we can think how to do it correctly. It should be governed in full by the mailbox queue, probably, to avoid races.